"Alert" component - actions implementation

didoo opened the pull request:
### :pushpin: Summary

This is an initial attempt to add support for the `actions` feature in the `Alert` component. 
I have implemented two ways to add actions to an alert:
- using the `@actions` argument (an array of objects) over which to loop and insert the corresponding HDS component
- a generic `<:actions>` named block that can be used to add any content into the "action" area

The reasons for having both is because we want to provide two ways for the users to declare the actions, a "controlled" one (that automatically sets the layout and possibly the `size` of the actions and maybe other properties by default) and an "escape hatch" in case they need to do something different/custom.

👉 **Notice**: at the moment I am not sure the `@actions` prop can do exactly what I have in mind (eg. does the `hash` work also for functions/callbacks? how do we pass extra attributes that then can be _splashed_ on the element? etc). Also, I am not sure the consumers will like this approach (it's a pattern more common/standard in React, where you can pass and spread props more easily than in Ember), they may prefer to just yield the action elements via the named block (but this in this case doesn't allow us to control spacing, layout, etc). See thread below for further discussion.

***

### 👀 How to review

👉 Review by files changed

Reviewer's checklist:

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201994286157147

didoo commented:
@amyrlam have a look at the code above. tomorrow we can discuss it together (I'll set up a meeting)

amyrlam commented on packages/components/addon/components/hds/alert/index.hbs:
```diff
@@ -11,13 +11,72 @@
 
-    {{#if @description}}
-      <div class="hds-alert__description">{{@description}}</div>
+    {{#if @actions}}
```
I haven't looked at any use cases in the "real world" of consumers yet, but I think that this is a really interesting approach! I can't recall seeing an Ember component that strongly recommends a default, but allows an escape hatch, in this manner

didoo commented:
## Some thoughts about the proposed patterns

### The problem we're trying to solve

According to the design specifications for the `Alert` component(s), the `Button[secondary|tertiary]` and the `Link(To)Standalone` component are allowed (as general guideline) as actions **only with the `small` size**:

<img width="400" alt="screenshot_1194" src="https://user-images.githubusercontent.com/686239/160405249-373e1c27-5734-41d3-ad12-9d3e9c7ffaff.png">

Also, there is no pre-defined list of actions: in some case we have none, in some cases only a `secondary` button, in other cases a `secondary` and a `tertiary`, in other cases a `secondary` and a `linkstandalone`.

So the question (problem) is: **_how do we allow the consumers to specify which buttons/links to use as actions in the `Alert` component, and possibly in doing so limit the elements to use only the correct size, while at the same time providing some forms of escape hatch in cases product designers need to "break the rules" for very specific use cases, contexts, exceptions?_**

Below I try to explain what are the possible options, what are the pros/cons and my thoughts on them.

### Option 1 - Just yield the actions in a `named block`

We could define the `Alert` component template in this way
```
<div class="hds-alert" ...attributes>
  [... other code here]
  {{#if (has-block "actions")}}
    <div class="hds-alert__actions">{{yield to="actions"}}</div>
  {{/if}}
</div>
```
so that the invocation would be like this:
```
  <Hds::Alert>
    <:actions>
      <Hds::Button @text="My button" @size="small" @color="secondary" />
      <Hds::Link::Standalone @icon="plus" @text="My link" href="#" @size="small" @color="secondary" />
    </:actions>
  </Hds::Alert>
```

**PROs:**
- it simple, it works, allows developers to deviate from the standards when needed

**CONs**:
- allows developers to deviate from the standards, so nothing prevents them from doing errors and using `medium` sized buttons in normal cases, or `primary` color for the button
- allows developers to add elements which are normally not supposed to go in the `action` area of the alert, so unintentional mistakes may happen (they have no indication of what components are supposed to go there, unless they look into the documentation)
- this solution does not allow us any form of control, by default

### Option 2 - Just yield the actions in a `named block` as pre-defined components

We could define the `Alert` component template in this way
```
<div class="hds-alert" ...attributes>
  [... other code here]
  {{#if (has-block "actions")}}
    <div class="hds-alert__actions">{{yield
          (hash
            Button=(component "hds/button")
            Link::Standalone=(component "hds/link/standalone")
            LinkTo::Standalone=(component "hds/link-to/standalone")
          )
    to="actions"}}</div>
  {{/if}}
</div>
```
so that the invocation would be like this:
```
  <Hds::Alert>
     <:actions as |A|>
      <A:Button @text="My button" @size="small" @color="secondary" />
      <A:Link::Standalone @icon="plus" @text="My link" href="#" @size="small" @color="secondary" />
    </:actions>
  </Hds::Alert>
```

**PROs:**
- TODO

**CONs**:
- TODO

### Option 3 - TODO

jesdavpet commented:
✅  I'd like to cast a vote in preference of **Option 1 - Just yield in a named block called actions** from the perspective of an HDS consumer. Let's just keep it simple, and flexible.

In terms of the cited **CONS:** listed for that Option 1:
> - _"allows developers to deviate from the standards ..."_
> - _"allows developers to add elements ..."_
> - _"this solution does not allow us any form of control, by default"_

Looking at this holistically, my inference is that these are motivated by a desire to try to govern designs. Is that correct? I think that's important consideration for consistency, and encouraging best practices.

I'd argue that it'd be much better to catch heuristic misses during design review, rather than through a restrictive Ember component API in code. I've observed UI engineers generally implement whatever's in Figma mocks, **very** faithfully _(often to the pixel)_.

Given that the Figma mock is being faithfully implemented: IMO setting up UI engineers to "police" design inconsistencies _after_ a design has been approved and handed off sounds like it could cause friction, and be potentially wasteful of time.

🧂 Grain of salt ... these are opinions.

venusang commented:
I vote for Option 1.  

Also, I am curious about Option 3.  It looks like Option 3 limits the order that the actions render.  Is this intentional?

meirish commented:
Just gonna chime in here and repeat what others said: FE more often than not builds what design specs out, so if things are consistent across designs, there shouldn't be a worry about what gets built. That mixed with the likelihood of users copy and pasting from docs (which I think is _very_ likely) I think the instances of "mis-use" are likely to be small.

I'd think that option 1 is still possible if option 2 is implemented (there's nothing forcing consumers to use the yielded components). So either of these is fine for me.

@didoo your question about passing default args when using the `component` helper - you can pass args, but not HTML attributes, and I believe the syntax is without the `@` - `(hash Button=(component "hds/button" size="small"))`.

didoo commented:
Thank you @jesdavpet @venusang @randallmorey @meirish for the precious feedback. I am closing this PR for now, and I'll implement option #2 in a separate PR so the history is clean (and we avoid the conflicts).

/cc @amyrlam

didoo commented:
### Update after #245 

While finalizing #245 we came around an issue related to whitespace, in particular on how Ember treats **whitespace** and **named blocks** and **yielded/contextual components**. I'll try to explain the issue here, what we learned trying to solve it, and how this affected the API implementation of `Alert` the component. The reason for adding all these learnings here is for future reference.

#### The problem

The initial implementation of the `Alert` component saw this code/API:
<img width="783" alt="screenshot_1375" src="https://user-images.githubusercontent.com/686239/167422483-43d9c3e8-728a-4843-8ebf-8f16ae05bf30.png">

When tested locally, in the showcase page, everything worked perfectly. But when trying to adopt the component in Cloud UI (see https://github.com/hashicorp/cloud-ui/pull/2429) we came across a problem with the whitespace. 

Essentially the invocation in Cloud UI looked like this:
<img width="749" alt="screenshot_1376" src="https://user-images.githubusercontent.com/686239/167422809-7e697159-dab4-436f-a8d0-4e5573f9d200.png">

When this alert was rendered, it had an extra visible space below the text because the `actions` block was rendered even if no contextual component was yielded to the `actions` block:
![toast-remove-user-success-after](https://user-images.githubusercontent.com/686239/167423103-c086b36d-ff3e-4be2-9082-0fe747205df6.png)

The reason for this is in the `if` declared **inside** the `<:actions>` named block invocation:
```
<:actions as |A|>
  {{#if (some condition)}}
    [some content]
  {{/if}}
</:actions>
```
At first sight this code makes totally sense, and one would not expect to see the block generated in the component (remember the condition `{{#if (has-block 'actions')}}`). In reality, Ember considers the whitespace before and after the IF **inside** the `<:actions>` block as **content**, so the check `{{#if (has-block 'actions')}}` returns true.

The first thought was to just invert the code in the invocation:
```
{{#if (some condition)}}
  <:actions as |A|>
    [some content]
  </:actions>
{{/if}}
```
and move the `{{if}}` check **outside** the `<:actions>` block, but in that case Ember throws an error:
<img width="618" alt="screenshot_1377" src="https://user-images.githubusercontent.com/686239/167424215-f2829d98-752b-4401-a91c-ad4eeca7d2c8.png">

The reason for this, even though counter-intuitive, is that the `<:actions>` is not a _tag_ but an _argument_ passed to the function/class of the component (see https://github.com/ember-polyfills/ember-named-blocks-polyfill/issues/25).

After spending almost one day trying to figure out this, I asked help in the [#tech-frontend Slack channel](https://hashicorp.slack.com/archives/C11JCBJTW/p1651069421716229) and I found out that there was a way in Ember to remove the extra whitespace using a specific syntax in Handlebars (`~` or "squishies"): https://handlebarsjs.com/guide/expressions.html#whitespace-control

So I added these _squishies_ in the component code, and this fixed the issue. Yay!!

But...

#### The "Toast" component

In the continuation of the work around #245 we have decided to split the `Alert` component, and have a dedicated `Toast` component only for that. The intention was to simply re-use the `Alert(Inline)` component, inside the `Toast` component (which at this point is nothing more than a simple wrapper around the `Alert`.

After a few back and forth and tests and [request for help (again) in the Slack channel ](https://hashicorp.slack.com/archives/C11JCBJTW/p1651583160749839) we came up with this nice API for the `Toast` component (this would avoid re-defining all the internals of `Alert` also inside the `Toast`):
<img width="254" alt="screenshot_1419" src="https://user-images.githubusercontent.com/686239/167427056-c6ab6881-2ad6-404e-ab5e-a797f918ee95.png">

But **again** the whitespace problem hit us! The space before/after the yield, inside the `<:actions>` was considered "content" and the "actions" block was rendered again, non empty! 😭 This time was even worse: for some reasons, even if we added the `~` squishies everywhere the block was rendered anyway. 

The reason for that (that's my understanding) is that what we're doing somehow is passing a "function" not just a block, or content, so the block is not undefined anyway. We tested also an approach following [what is done](https://github.com/prysmex/ember-eui/blob/master/packages/core/addon/components/eui-field-text/index.hbs) in this UI library code (https://github.com/prysmex/ember-eui) essentially defining a specific prop `ignoreAlertBlock` and pass it down to the `Alert` component, and at first sight seemed to work, but then when the `Toast` was invoked with an `{{#if}}` inside the whitespace problem appeared again.
<img width="1792" alt="modified" src="https://user-images.githubusercontent.com/686239/167428751-a14c5530-b3aa-49b9-bb2e-5b14cff5806b.png">

The only way, again, to solve the problem was not only to remove the whitespace from the `Toast`
<img width="594" alt="modified" src="https://user-images.githubusercontent.com/686239/167428899-c457d69a-bc99-4b95-ab00-9219ffa04952.png">
but (**much worse!**) we would have to ask to do the same for the consumers! Which is not feasible, it would be a really bad DX.

In the thread on Slack it was pointed out that this is a known issue, and no one has a real solution to the problem:
- https://github.com/emberjs/rfcs/issues/735#issuecomment-817983427
  - https://github.com/emberjs/rfcs/issues/735#issuecomment-1112284836
- https://github.com/emberjs/rfcs/pull/460#issuecomment-902961346

#### The "solution" 

At the end the decision I took was this:
- [use CSS to add the space to the `actions` children](https://hashicorp.slack.com/archives/C11JCBJTW/p1651747371587809?thread_ts=1651691482.276839&cid=C11JCBJTW) instead of the block itself (not great, but it works)
- remove the named blocks from the `Alert` component: they were causing too many issues, and in this way the API of the `Alert` component would follow the same pattern used for the `Dropdown` (direct declaration of yielded components)
  - with this it was also possible to remove all the `~` squishies from the code (the whitespace was not a problem anymore at this point, because there were no named blocks carrying it forward in their invocations.

The final API for the component looked like this at the end (we also added a `generic` container, as yielded component):
```
  <Hds::Alert @type="inline" @title="Title here" @description="Description here" as |A|>
    <A.Button @text="Your action" @color="secondary" />
    <A.LinkTo::Standalone @color="secondary" @icon="plus" @text="Another action" @route="index" />
    <A.Link::Standalone @icon="arrow-right" @iconPosition="trailing" @text="Another action" href="#" />
    <A.Generic>[some content here]</A.Generic>
  </Hds::Alert>
```
